roofit/hs3/src/RooJSONFactoryWSTool.cxx

```diff
@@ -0,0 +1,1432 @@
+               v->setVal(point[j].val_float());
+            }
+            data->add(vars, weights[i].val_float());
+         }
+         dataMap[name] = data;
+      } else if (p.has_child("index")) {
+         // combined measurement
+         auto subMap = loadData(p);
+         auto catname = p["index"].val();
+         RooCategory *channelCat = _workspace->cat(catname.c_str());
+         if (!channelCat) {
+            coutE(InputArguments) << "RooJSONFactoryWSTool(" << GetName() << ") failed to retrieve channel category "
+                                  << catname << std::endl;
+         } else {
+            RooArgSet allVars;
+            allVars.add(*channelCat);
+            std::map<std::string, RooDataSet *> datasets;
+            for (const auto &subd : subMap) {
+               allVars.add(*subd.second->get());
+               if (subd.second->InheritsFrom(RooDataHist::Class())) {
+                  datasets[subd.first] = unbinned(((RooDataHist *)subd.second));
+               } else {
+                  datasets[subd.first] = ((RooDataSet *)subd.second);
+               }
+            }
+            RooRealVar *weightVar = this->getWeightVar("weight");
+            allVars.add(*weightVar);
+            RooDataSet *data = new RooDataSet(name.c_str(), name.c_str(), allVars, RooFit::Index(*channelCat),
+                                              RooFit::Import(datasets), RooFit::WeightVar(*weightVar));
+            dataMap[name] = data;
+         }
+      } else {
+         coutE(InputArguments) << "RooJSONFactoryWSTool(" << GetName() << ") failed to create dataset" << name
+                               << std::endl;
+      }
+   }
+   return dataMap;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+// exporting data
+void RooJSONFactoryWSTool::exportData(RooAbsData *data, JSONNode &n)
+{
+   RooArgSet observables(*data->get());
+   auto &output = n[data->GetName()];
+   output.set_map();
+
+   // find category observables
+   RooAbsCategory *cat = NULL;
+   for (const auto &obs : observables) {
+      if (obs->InheritsFrom(RooAbsCategory::Class())) {
+         if (cat) {
+            RooJSONFactoryWSTool::error("dataset '" + std::string(data->GetName()) +
+                                        " has several category observables!");
+         } else {
+            cat = (RooAbsCategory *)(obs);
+         }
+      }
+   }
+
+   if (cat) {
+      // this is a composite dataset
+      RooDataSet *ds = (RooDataSet *)(data);
+      output["index"] << cat->GetName();
+      TList *dataList = ds->split(*(cat), true);
+      if (!dataList) {
+         RooJSONFactoryWSTool::error("unable to split dataset '" + std::string(ds->GetName()) + "' at '" +
+                                     std::string(cat->GetName()) + "'");
+      }
+      for (const auto &d : *dataList) {
+         this->exportData((RooAbsData *)d, output);
+      }
```

Comment:
```suggestion
       for (RooAbsData * absData : static_range_cast<RooAbsData*>(*dataList)) {
          this->exportData(absData, output);
       }
```
While fixing the shadowing warning (see CI), you can also use the new `static_range_cast`.